Guled/add comic book network request

Somnibyte opened the pull request:
* Comics can now be added to the database
* ComicFormViewController now uses RxSwift.

Commit:
Add S3ImageURL struct. Add AddComicBook and ComicPhotoURL endpoints.

Commit:
Implement Rx for UITextFields, UIPickerView, and UIButtons.

Commit:
Cleanup syntax, Add disposeBag to subscribe methods.

Remove self.

Avoid force unwrap. Set rootViewController back to RootViewController. Empty authorization header.

Rename S3ImageURL to AmazonS3ImageUrl for clarity.

Reformat closures.

IntrepidBuildServer commented:
<!--
  0 Errors
  0 Warnings
  2 Messages: Executed 2 tests, with 0 failu...
  0 Markdowns
-->
<table>
  <thead>
    <tr>
      <th width="50"></th>
      <th width="100%" data-danger-table="true" data-kind="Message">
          2 Messages
      </th>
     </tr>
  </thead>
  <tbody>
    <tr>
      <td>:book:</td>
      <td data-sticky="false">Executed 2 tests, with 0 failures (0 unexpected) in 0.428 (0.431) seconds</td>
    </tr>
    <tr>
      <td>:book:</td>
      <td data-sticky="false">Executed 1 test, with 0 failures (0 unexpected) in 15.048 (15.051) seconds</td>
    </tr>
  </tbody>
</table>

<p align="right" data-meta="generated_by_danger">
  Generated by :no_entry_sign: <a href="http://danger.systems/">Danger</a>
</p>

mayasaxena commented on lucien-ios/APIClient/LucienAPIClient.swift:
```diff
@@ -93,4 +93,59 @@ final class LucienAPIClient: APIClient {
+                }
+                let decoder = JSONDecoder()
+                decoder.dateDecodingStrategy = .iso8601
+                guard let s3ImageURL = try? decoder.decode(AmazonS3ImageURL.self, from: result) else {
```
In this case, rather than using the optional `try`, wrap the decode call in a `do-catch` and in the `catch` block, call the completion with the error returned instead of `cannotDecode`. Take a look at `getCurrentUser` for an example

mayasaxena commented on lucien-ios/APIClient/LucienAPIClient.swift:
```diff
@@ -93,4 +93,59 @@ final class LucienAPIClient: APIClient {
+                    return completion(.failure(LucienAPIError.noResult))
+                }
+                let decoder = JSONDecoder()
+                decoder.dateDecodingStrategy = .iso8601
```
Do you need this here? Do you know what it's doing?

Somnibyte replied:
I don't need it. That was a mistake.

mayasaxena commented on lucien-ios/APIClient/LucienAPIClient.swift:
```diff
@@ -93,4 +93,59 @@ final class LucienAPIClient: APIClient {
+        sendRequest(urlRequest) { response in
+            print(response)
+            switch response {
+            case .success(_):
```
Remove the `(_)` for now, you can put it back in with the comic at a later date. Rationale: even code that will eventually be updated should still adhere to the style guide since priorities may change and the update may not happen.

mayasaxena commented on lucien-ios/APIClient/LucienRequest.swift:
```diff
@@ -17,6 +17,17 @@ enum LucienRequest: Request {
     case getDashboard
     case hasCollection
+    case createPhotoURL
+    case addComicBook(comicTitle: String,
```
Instead of passing all of this in, pass a `Comic` and we can get the values from there

mayasaxena commented on lucien-ios/APIClient/LucienRequest.swift:
```diff
@@ -74,12 +97,34 @@ enum LucienRequest: Request {
+                           let returnDate,
+                           let condition,
+                           let genre):
+            return ["comic": ["comic_title": comicTitle,
```
Instead of creating this on the fly, it should be a property on `Comic` and I suspect we can use `Codable` to generate this for us

mayasaxena commented on lucien-ios/AddAndEditNewComicBook/ComicFormViewController.swift:
```diff
@@ -72,29 +77,65 @@ final class ComicFormViewController: UIViewController, AlertDisplaying {
+            self?.seriesTitleTextField.borderColor = LucienTheme.dark
+            self?.checkIfAllRequiredFieldsAreFilled()
+        } >>> disposeBag
+        storyTitleTextField.rx.controlEvent([.editingChanged]).asObservable().subscribeNext { [weak self] in
```
I would add newlines after closures

mayasaxena commented on lucien-ios/AddAndEditNewComicBook/ComicFormViewController.swift:
```diff
@@ -72,29 +77,65 @@ final class ComicFormViewController: UIViewController, AlertDisplaying {
+            self?.comicFormViewModel.genre = selectedGenre
+            self?.selectAGenreButton.setTitle(selectedGenre?.title, for: .normal)
+        } >>> disposeBag
+        conditionPicker.rx.itemSelected.subscribeNext { [weak self] row, _ in
```
Newline before this

mayasaxena commented on lucien-ios/AddAndEditNewComicBook/ComicFormViewController.swift:
```diff
@@ -72,29 +77,65 @@ final class ComicFormViewController: UIViewController, AlertDisplaying {
-            issueTextField.text = comicFormViewModel.issue ?? ""
-            publisherTextField.text = comicFormViewModel.publisher ?? ""
-            releaseDateTextField.text = comicFormViewModel.release ?? ""
+            seriesTitleTextField.text = comicFormViewModel.seriesTitle.value
```
You shouldn't need to do this anymore - the bidirectional binding should fill it for you

mayasaxena commented on lucien-ios/AddAndEditNewComicBook/ComicFormViewModel.swift:
```diff
@@ -16,17 +19,23 @@ enum  ComicFormMode {
-    var issue: String?
-    var publisher: String?
-    var release: String?
+    var seriesTitle = Variable<String>("")
```
If you initialize this with an initial value (the '("")' part, you don't need to explicitly declare it's a variable of Strings. Also this can be a let. Aka you can just say this `let seriesTitle = Variable("")`. Same for the ones below

mayasaxena commented on lucien-ios/LoginScreen/LoginViewModel.swift:
```diff
@@ -49,4 +49,74 @@ final class LoginViewModel {
+                      genre: String?,
+                      completion: @escaping (Error?) -> Void) {
+
+        lucienAPIClient.addComicBook(comicTitle: comicTitle,
```
All parameters here should be on a new line, indented by one from `lucienAPIClient`

mayasaxena commented on lucien-ios/LoginScreen/LoginViewModel.swift:
```diff
@@ -49,4 +49,74 @@ final class LoginViewModel {
+                                     comicPhotoURL: comicPhotoURL,
+                                     returnDate: returnDate,
+                                     condition: condition,
+                                     genre: genre) { response in
```
the `) { response in` should be on its own line indented at the same level as `lucienAPIClient`

mayasaxena commented on lucien-ios/LoginScreen/LoginViewModel.swift:
```diff
@@ -49,4 +49,74 @@ final class LoginViewModel {
+                                     returnDate: returnDate,
+                                     condition: condition,
+                                     genre: genre) { response in
+                                    switch response {
```
These should all be moved almost all the way to the left

mayasaxena commented on lucien-ios/LoginScreen/LoginViewModel.swift:
```diff
@@ -49,4 +49,74 @@ final class LoginViewModel {
+        }
+    }
+
+    func sendRequestToS3(data: Data, urlString: String, completion: @escaping (Error?) -> Void) {
```
This should be moved into the API client

mayasaxena commented on lucien-ios/LoginScreen/LoginViewModel.swift:
```diff
@@ -49,4 +49,74 @@ final class LoginViewModel {
+        }
+    }
+
+    func createPhotoURL(image: UIImage?, completion: @escaping (String) -> Void) {
```
This naming is misleading - you're not just creating the URL, you're also uploading the image, so it should be named accordingly

mayasaxena commented on lucien-ios/LoginScreen/LoginViewModel.swift:
```diff
@@ -49,4 +49,74 @@ final class LoginViewModel {
+        task.resume()
+    }
+
+    func upload(image: UIImage, urlString: String, completion: @escaping (Error?) -> Void) {
```
This should be private

Commit:
Impelment feedback.

Send requests to server to add comics. Refactor ComicFormViewModel to incorporate Rxswift.

Remove self. self was never used.

Remove unwanted space.

Add genre to Comic struct init.

Add contants to LucienConstants file. Fix formatting of >>> disposeBag lines.

Rename long variable names. Fix formatting of LucienConstant code.

IntrepidBuildServer commented:
<!--
  0 Errors
  1 Warning: <a href="https://github.com/In...
  2 Messages: Executed 2 tests, with 0 failu...
  0 Markdowns
-->
<table>
  <thead>
    <tr>
      <th width="50"></th>
      <th width="100%" data-danger-table="true" data-kind="Warning">
          1 Warning
      </th>
     </tr>
  </thead>
  <tbody>
    <tr>
      <td>:warning:</td>
      <td data-sticky="false"><a href="https://github.com/IntrepidPursuits/lucien-ios/blob/5bb97de47b88cdb6af40bbd50da8251b495b7f86/ComicFormViewModel.swift#L140">ComicFormViewModel.swift#L140</a> - <strong><a href="https://github.com/IntrepidPursuits/lucien-ios/blob/5bb97de47b88cdb6af40bbd50da8251b495b7f86/lucien-ios/AddAndEditNewComicBook/ComicFormViewModel.swift#L140">lucien-ios/AddAndEditNewComicBook/ComicFormViewModel.swift#L140</a></strong>: variable ‘self’ was written to, but never read  <br /><code>
        lucienAPIClient.addComicBook(comic: comic) { [weak self]  response in
</code></td>
    </tr>
  </tbody>
</table>
<table>
  <thead>
    <tr>
      <th width="50"></th>
      <th width="100%" data-danger-table="true" data-kind="Message">
          2 Messages
      </th>
     </tr>
  </thead>
  <tbody>
    <tr>
      <td>:book:</td>
      <td data-sticky="false">Executed 2 tests, with 0 failures (0 unexpected) in 0.402 (0.407) seconds</td>
    </tr>
    <tr>
      <td>:book:</td>
      <td data-sticky="false">Executed 1 test, with 0 failures (0 unexpected) in 8.610 (8.612) seconds</td>
    </tr>
  </tbody>
</table>

<p align="right" data-meta="generated_by_danger">
  Generated by :no_entry_sign: <a href="http://danger.systems/">Danger</a>
</p>

IntrepidBuildServer commented:
<!--
  1 Error: fatal error: file ‘/Users/jenk...
  0 Warnings
  2 Messages: Executed 2 tests, with 0 failu...
  0 Markdowns
-->
<table>
  <thead>
    <tr>
      <th width="50"></th>
      <th width="100%" data-danger-table="true" data-kind="Error">
          1 Error
      </th>
     </tr>
  </thead>
  <tbody>
    <tr>
      <td>:no_entry_sign:</td>
      <td data-sticky="false">fatal error: file ‘/Users/jenkins/Jenkins/workspace/s-pipeline_lucien-ios_PR-18-Y5PVQN6XJGBUHWSNFNVZWDNICZHAGWKNIPJLHI2OZY3JWLJZXJEA/Pods/Google/Headers/module.modulemap’ has been modified since the precompiled header ‘/Users/jenkins/Jenkins/workspace/s-pipeline_lucien-ios_PR-18-Y5PVQN6XJGBUHWSNFNVZWDNICZHAGWKNIPJLHI2OZY3JWLJZXJEA/DerivedData/Build/Intermediates.noindex/PrecompiledHeaders/lucien-ios-Bridging-Header-swift_3AARK3WWC8FK5-clang_2O9YIEWVALUCS.pch’ was built</td>
    </tr>
  </tbody>
</table>
<table>
  <thead>
    <tr>
      <th width="50"></th>
      <th width="100%" data-danger-table="true" data-kind="Message">
          2 Messages
      </th>
     </tr>
  </thead>
  <tbody>
    <tr>
      <td>:book:</td>
      <td data-sticky="false">Executed 2 tests, with 0 failures (0 unexpected) in 0.350 (0.352) seconds</td>
    </tr>
    <tr>
      <td>:book:</td>
      <td data-sticky="false">Executed 1 test, with 0 failures (0 unexpected) in 18.569 (18.572) seconds</td>
    </tr>
  </tbody>
</table>

<p align="right" data-meta="generated_by_danger">
  Generated by :no_entry_sign: <a href="http://danger.systems/">Danger</a>
</p>

IntrepidBuildServer commented:
<!--
  0 Errors
  0 Warnings
  2 Messages: Executed 2 tests, with 0 failu...
  0 Markdowns
-->
<table>
  <thead>
    <tr>
      <th width="50"></th>
      <th width="100%" data-danger-table="true" data-kind="Message">
          2 Messages
      </th>
     </tr>
  </thead>
  <tbody>
    <tr>
      <td>:book:</td>
      <td data-sticky="false">Executed 2 tests, with 0 failures (0 unexpected) in 0.391 (0.393) seconds</td>
    </tr>
    <tr>
      <td>:book:</td>
      <td data-sticky="false">Executed 1 test, with 0 failures (0 unexpected) in 13.099 (13.101) seconds</td>
    </tr>
  </tbody>
</table>

<p align="right" data-meta="generated_by_danger">
  Generated by :no_entry_sign: <a href="http://danger.systems/">Danger</a>
</p>

mayasaxena commented on lucien-ios/AddAndEditNewComicBook/ComicFormViewController.swift:
```diff
@@ -72,39 +82,75 @@ final class ComicFormViewController: UIViewController, AlertDisplaying {
+
+        releaseDateTextField.rx.controlEvent([.editingChanged]).asObservable().subscribeNext { [weak self] in
+            guard let text = self?.releaseDateTextField.text else { return }
+            if text.characters.count > 4 {
```
Store this as a constant in a private struct in this class called `Constants`

mayasaxena commented on lucien-ios/AddAndEditNewComicBook/ComicFormViewController.swift:
```diff
@@ -122,12 +168,22 @@ final class ComicFormViewController: UIViewController, AlertDisplaying {
+            }
+            let cancelAction = UIAlertAction(title: "Cancel", style: .cancel)
+            self?.showAlert(title: "", message: "This will delete your current comic information.", actions: [goBackAction, cancelAction], preferredStyle: .actionSheet)
+            } >>> disposeBag
```
Un-indent this one level so it matches `backButton`

mayasaxena commented on lucien-ios/AddAndEditNewComicBook/ComicFormViewController.swift:
```diff
@@ -137,13 +193,17 @@ final class ComicFormViewController: UIViewController, AlertDisplaying {
+                    self?.showAlert(title: "Error", message: "Our service is currently encountering an issue. Please ensure that you are connected to the internet and try again.")
+                }
+            }
+            } >>> disposeBag
```
Un-indent this one level so it matches `finishButton`

mayasaxena commented on lucien-ios/AddAndEditNewComicBook/ComicFormViewController.swift:
```diff
@@ -153,7 +213,13 @@ final class ComicFormViewController: UIViewController, AlertDisplaying {
+        doneButton.rx.tap.subscribeNext { [weak self] in
+            self?.view.endEditing(true)
+            self?.deregisterFromKeyboardNotifications()
+            } >>> disposeBag
```
unindent

mayasaxena commented on lucien-ios/AddAndEditNewComicBook/ComicFormViewModel.swift:
```diff
@@ -0,0 +1,171 @@
+    case edit
+}
+
+/// The comic attribute of the ComicFormViewModel is an Observable that attempts to combine the
```
This comment is too verbose. Most of the information within it can be inferred by seeing the use of the struct. It's enough to just say something like "encapsulates optional `Comic` properties to circumvent 8 parameter limit of `Observable.combineLatest`. Rule of thumb: if you have to write a comment that's more than two sentences long to explain something that's not domain-specific (a math formula etc), either the comment is too long or your object is badly named/designed.

mayasaxena commented on lucien-ios/AddAndEditNewComicBook/ComicFormViewModel.swift:
```diff
@@ -0,0 +1,171 @@
+
+    // MARK: - Instance Methods
+
+    /// Called when the finished button, within the ComicFormViewController, is tapped.
```
This comment doesn't add any information, so it should be removed

mayasaxena commented on lucien-ios/AddAndEditNewComicBook/ComicFormViewModel.swift:
```diff
@@ -0,0 +1,171 @@
+    }
+
+    private func upload(image: UIImage, urlString: String, completion: @escaping (Error?) -> Void) {
+        guard let imageData = UIImageJPEGRepresentation(image, 0.9) else { return }
```
0.9 should be extracted out into a constant

IntrepidBuildServer commented:
<!--
  1 Error: fatal error: file ‘/Users/jenk...
  0 Warnings
  2 Messages: Executed 2 tests, with 0 failu...
  0 Markdowns
-->
<table>
  <thead>
    <tr>
      <th width="50"></th>
      <th width="100%" data-danger-table="true" data-kind="Error">
          1 Error
      </th>
     </tr>
  </thead>
  <tbody>
    <tr>
      <td>:no_entry_sign:</td>
      <td data-sticky="false">fatal error: file ‘/Users/jenkins/Jenkins/workspace/s-pipeline_lucien-ios_PR-18-Y5PVQN6XJGBUHWSNFNVZWDNICZHAGWKNIPJLHI2OZY3JWLJZXJEA/Pods/Google/Headers/module.modulemap’ has been modified since the precompiled header ‘/Users/jenkins/Jenkins/workspace/s-pipeline_lucien-ios_PR-18-Y5PVQN6XJGBUHWSNFNVZWDNICZHAGWKNIPJLHI2OZY3JWLJZXJEA/DerivedData/Build/Intermediates.noindex/PrecompiledHeaders/lucien-ios-Bridging-Header-swift_3AARK3WWC8FK5-clang_2O9YIEWVALUCS.pch’ was built</td>
    </tr>
  </tbody>
</table>
<table>
  <thead>
    <tr>
      <th width="50"></th>
      <th width="100%" data-danger-table="true" data-kind="Message">
          2 Messages
      </th>
     </tr>
  </thead>
  <tbody>
    <tr>
      <td>:book:</td>
      <td data-sticky="false">Executed 2 tests, with 0 failures (0 unexpected) in 0.406 (0.409) seconds</td>
    </tr>
    <tr>
      <td>:book:</td>
      <td data-sticky="false">Executed 1 test, with 0 failures (0 unexpected) in 8.250 (8.253) seconds</td>
    </tr>
  </tbody>
</table>

<p align="right" data-meta="generated_by_danger">
  Generated by :no_entry_sign: <a href="http://danger.systems/">Danger</a>
</p>

mayasaxena reviewed (changes requested):
Last round of comments

IntrepidBuildServer commented on lucien-ios/LucienConstants.swift:
```diff
@@ -34,6 +39,18 @@ final class LucienConstants {
+    static let coverPhotoTopTitleEdgeInset =  CGFloat(25)
+
+
+
```
![MINOR](https://sonarsource.github.io/sonar-github/severity-minor.png 'Severity: MINOR') Vertical Whitespace Violation: Limit vertical whitespace to a single empty line. Currently 3. [![rule](https://sonarsource.github.io/sonar-github/rule.png)](https://sonar.intrepid.io/coding_rules#rule_key=SwiftLint%3Avertical_whitespace)

IntrepidBuildServer commented on lucien-ios/LucienConstants.swift:
```diff
@@ -43,3 +60,4 @@ final class LucienConstants {
     static let overlayButtonTopConstraint = CGFloat(10)
     static let overlayButtonLeadingConstraint = CGFloat(15)
 }
+
```
![MINOR](https://sonarsource.github.io/sonar-github/severity-minor.png 'Severity: MINOR') Trailing Newline Violation: Files should have a single trailing newline. [![rule](https://sonarsource.github.io/sonar-github/rule.png)](https://sonar.intrepid.io/coding_rules#rule_key=SwiftLint%3Atrailing_newline)

IntrepidBuildServer commented:
SonarQube analysis reported 2 issues
* ![MINOR][MINOR] 2 minor

Watch the comments in this conversation to review them.

[MINOR]: https://sonarsource.github.io/sonar-github/severity-minor.png 'Severity: MINOR'

IntrepidBuildServer commented:
<!--
  0 Errors
  0 Warnings
  4 Messages: Executed 2 tests, with 0 failu...
  0 Markdowns
-->
<table>
  <thead>
    <tr>
      <th width="50"></th>
      <th width="100%" data-danger-table="true" data-kind="Message">
          4 Messages
      </th>
     </tr>
  </thead>
  <tbody>
    <tr>
      <td>:book:</td>
      <td data-sticky="false">Executed 2 tests, with 0 failures (0 unexpected) in 0.377 (0.379) seconds</td>
    </tr>
    <tr>
      <td>:book:</td>
      <td data-sticky="false">Executed 1 test, with 0 failures (0 unexpected) in 7.485 (7.487) seconds</td>
    </tr>
    <tr>
      <td>:book:</td>
      <td data-sticky="false">Executed 2 tests, with 0 failures (0 unexpected) in 0.377 (0.379) seconds</td>
    </tr>
    <tr>
      <td>:book:</td>
      <td data-sticky="false">Executed 1 test, with 0 failures (0 unexpected) in 7.485 (7.487) seconds</td>
    </tr>
  </tbody>
</table>

<p align="right" data-meta="generated_by_danger">
  Generated by :no_entry_sign: <a href="http://danger.systems/">Danger</a>
</p>

ghost commented:
<!--
  1 Error: fatal error: file ‘/Users/jenk...
  0 Warnings
  2 Messages: Executed 2 tests, with 0 failu...
  0 Markdowns
-->
<table>
  <thead>
    <tr>
      <th width="50"></th>
      <th width="100%" data-danger-table="true" data-kind="Error">
          1 Error
      </th>
     </tr>
  </thead>
  <tbody>
    <tr>
      <td>:no_entry_sign:</td>
      <td data-sticky="false">fatal error: file ‘/Users/jenkins/Jenkins/workspace/s-pipeline_lucien-ios_PR-18-Y5PVQN6XJGBUHWSNFNVZWDNICZHAGWKNIPJLHI2OZY3JWLJZXJEA/Pods/Google/Headers/module.modulemap’ has been modified since the precompiled header ‘/Users/jenkins/Jenkins/workspace/s-pipeline_lucien-ios_PR-18-Y5PVQN6XJGBUHWSNFNVZWDNICZHAGWKNIPJLHI2OZY3JWLJZXJEA/DerivedData/Build/Intermediates.noindex/PrecompiledHeaders/lucien-ios-Bridging-Header-swift_3AARK3WWC8FK5-clang_2O9YIEWVALUCS.pch’ was built</td>
    </tr>
  </tbody>
</table>
<table>
  <thead>
    <tr>
      <th width="50"></th>
      <th width="100%" data-danger-table="true" data-kind="Message">
          2 Messages
      </th>
     </tr>
  </thead>
  <tbody>
    <tr>
      <td>:book:</td>
      <td data-sticky="false">Executed 2 tests, with 0 failures (0 unexpected) in 0.546 (0.549) seconds</td>
    </tr>
    <tr>
      <td>:book:</td>
      <td data-sticky="false">Executed 1 test, with 0 failures (0 unexpected) in 9.031 (9.034) seconds</td>
    </tr>
  </tbody>
</table>

<p align="right" data-meta="generated_by_danger">
  Generated by :no_entry_sign: <a href="http://danger.systems/">Danger</a>
</p>
